keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,789 @@
+scale. Now I want to propagate topologically-aware service endpoints (local,
+regional, global) to all clusters, so that other services in my clusters can
+access instances of this service in priority order based on availability and
+locality. Requests to my replicated service should seamlessly transition (within
+SLO for dropped requests) between instances of my service in case of failure or
+removal without action by or impact on the caller. Routing to my replicated
+service should optimize for cost metric (e.g.prioritize traffic local to zone,
+region).
+
+```
+<<[UNRESOLVED]>>
+Due to additional constraints that apply to stateful services (e.g. each cluster
+potentially having pods with the conflicting hostnames `set-name-0`, `set-name-1`,
+etc.) we are only targeting stateless services for the multi-cluster backed use
+case for now.
+<<[/UNRESOLVED]>>
+```
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+### Exporting Services
+
+Services will not be visible to other clusters in the supercluster by default.
+They must be explicitly marked for export by the user. This allows users to
+decide exactly which services should be visible outside of the local cluster.
+
+Tooling may (and likely will, in the future) be built on top of this to simplify
+the user experience. Some initial ideas are to allow users to specify that all
+services in a given namespace or in a namespace selector or even a whole cluster
+should be automatically exported by default. In that case, a `ServiceExport`
+could be automatically created for all `Services`. This tooling will be designed
+in a separate doc, and is secondary to the main API proposed here.
+
+To mark a service for export to the supercluster, a user will create a
+ServiceExport CR:
+
+```golang
+// ServiceExport declares that the associated service should be exported to
+// other clusters.
+type ServiceExport struct {
+        metav1.TypeMeta `json:",inline"`
+        // +optional
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        // +optional
```

Comment:
I'm wondering if we don't want to explicitly add Spen and store ObjectReference to an underlying service there.

It's probably a bit of information duplication for now (as we rely on same name/namespace), but it would make it more explicit (e.g. we were relying on this 1-1 mapping for Service-Endpoints and this bite us). Being more explicit in my opinion may appear to be beneficial over time.

Reply:
I added a section on this to the Alternatives section. We had discussed this approach, but it makes for fundamentally different behavior:
- OR allows cross namespace exports
- OR allows multiple exports for a given Service
- OR can't be blocked by RBAC

The above can all be addressed by the controller, but then we risk implementation dependent behavior.

Maybe it would make sense to put it in `Status` instead, and have the controller auto populate it based on the associated service? That would give tooling an explicit link, but avoid any questions about what happens when the OR is set to a non-name-mapped service. The OR would always be redundant information, but might make the association more clear - and help identify when the requested service doesn't exist.

Reply:
OK - I buy that. Thanks for adding the alternatives considered - that explanation is super important in my opinion.

Reply:
absolutely - anything that looks unusual should be well explained